lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -1092,9 +1096,223 @@ class DecomposeAtenMvOp : public OpRewritePattern<AtenMvOp> {
+    postPermuteShape.reserve(nLeadingDims + 5);
+
+    SmallVector<Value> outShape;
+    outShape.reserve(nLeadingDims + 3);
+
+    SmallVector<Value> permutation;
+    permutation.reserve(nLeadingDims + 5);
+
+    for (unsigned i = 0; i < nLeadingDims; ++i) {
+      auto dimensionAttr = rewriter.getI64IntegerAttr(i);
+      Value dimensionValue = rewriter.create<ConstantIntOp>(loc, dimensionAttr);
+      Value leadingDimSize =
+          rewriter.createOrFold<AtenSizeIntOp>(loc, inValue, dimensionValue);
+      prePermuteShape.push_back(leadingDimSize);
+      postPermuteShape.push_back(leadingDimSize);
+      outShape.push_back(leadingDimSize);
+      permutation.push_back(dimensionValue);
+
+    }
+
+    const auto inOptionalDType = inType.getOptionalDtype();
+
+    auto getTypeFromShape = [inOptionalDType](auto &&vals) {
+      // Get a vector of integers from a vector of Values.
+      auto getIntShape = [](auto &&vals) {
+        SmallVector<int64_t> shape;
+        shape.reserve(vals.size());
+        for (auto v : vals) {
+          int64_t cst_val;
+          if (matchPattern(v, m_TorchConstantInt(&cst_val))) {
+            shape.push_back(cst_val);
+          } else {
+            shape.push_back(kUnknownSize);
+          }
+        }
+        return shape;
+      };
+
+      const auto intShape = getIntShape(vals);
+      return ValueTensorType::get(vals[0].getContext(),
+                                  llvm::ArrayRef(intShape), inOptionalDType);
+    };
+
+    prePermuteShape.insert(prePermuteShape.end(),
+                           {outC, factor, factor, inH, inW});
+
+    postPermuteShape.insert(postPermuteShape.end(),
+                            {outC, inH, factor, inW, factor});
+
+    outShape.insert(outShape.end(), {outC, outH, outW});
+
+    SmallVector<uint64_t> permutationTail{0, 3, 1, 4, 2};
+    for (uint64_t d : permutationTail) {
+      permutation.push_back(rewriter.create<ConstantIntOp>(
+          loc, rewriter.getI64IntegerAttr(nLeadingDims + d)));
+    }
+
+    auto listType = Torch::ListType::get(Torch::IntType::get(op.getContext()));
+
+
+     Value shapeA =
+         rewriter.create<PrimListConstructOp>(loc, listType, prePermuteShape);
+
+     Value A = rewriter.create<AtenReshapeOp>(
+         loc, getTypeFromShape(prePermuteShape), inValue, shapeA);
+
+     Value permuteDimsOrder = rewriter.create<PrimListConstructOp>(
+         loc, Torch::ListType::get(Torch::IntType::get(op->getContext())),
+         permutation);
+
+     Value B = rewriter.create<AtenPermuteOp>(
+         loc, getTypeFromShape(postPermuteShape), A, permuteDimsOrder);
+
+     Value outShapeList =
+         rewriter.create<PrimListConstructOp>(loc, listType, outShape);
+
+     // TODO(jn) figure out why the deduced return type (like
+     // !torch.vtensor<[3,4,6],si64>) cannot be used for the type of the
+     // replacement of the pixel_shuffle op's output. The pattern seems to
+     // require the generic !torch.vtensor type.
```

Comment:
I think the reason you need to cast back to an unranked tensor is because you are missing a shape conversion function here: https://github.com/llvm/torch-mlir/blob/main/projects/pt1/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py

There should be an entry for `aten.pixel_shuffle` in `include/torch-mlir/Dialect/Torch/IR/JITOperatorRegistryDump.txt` that has shape/dtype/(decomposition) signature functions that you can add. Here is an example for aten._fw_primal
```
  shape_function_signature = def aten〇_fw_primal〡shape(self: List[int], level: int) -> List[int]:
  decomposition_function_signature = def aten〇_fw_primal〡decomposition(self: Tensor, level: int) -> Tensor:
  dtype_function_signature = def aten〇_fw_primal〡dtype(self_rank_dtype: Tuple[int, int], level: int) -> int:

```

Reply:
Nice, I was wondering where to implement the type inference. That fixes the problem.